dagshub/streaming/filesystem.py

```diff
@@ -482,38 +469,95 @@ def scandir(self, path='.'):
+                    name = Path(f.path).name
                     if name not in local_filenames:
-                        yield dagshub_DirEntry(self, path / name, f['type'] == 'dir', is_binary=is_bytes_path_arg)
+                        yield dagshub_DirEntry(self, path / name, f.type == 'dir', is_binary=is_bytes_path_arg)
         else:
             for entry in self.__scandir(path):
                 yield entry
 
-    @cache_by_path
-    def _api_listdir(self, path: str, include_size: bool = False):
-        return self.http_get(f'{self.content_api_url}/{path}',
-                             params={'include_size': 'true'} if include_size else {},
-                             headers=config.requests_headers)
-
-    def _api_download_file_git(self, path: str):
-        return self.http_get(f'{self.raw_api_url}/{path}', headers=config.requests_headers, timeout=None)
+    def _get_special_paths(
+        self, dh_path: DagshubPath, relative_to: PathLike, is_binary: bool
+    ) -> Set["dagshub_DirEntry"]:
+        def generate_entry(path, is_directory):
+            if isinstance(path, str):
+                path = Path(path)
+            return dagshub_DirEntry(self, relative_to / path,
+                                    is_directory=is_directory, is_binary=is_binary)
+
+        has_storages = len(self._storages) > 0
+        res = set()
+        str_path = dh_path.relative_path.as_posix()
+        if str_path == ".":
+            res.add(generate_entry(SPECIAL_FILE, False))
+            if has_storages:
+                res.add(generate_entry(".dagshub", True))
+        elif str_path.startswith(".dagshub") and has_storages:
+            storage_paths = [s.path_in_mount for s in self._storages]
+            for sp in storage_paths:
+                try:
+                    relpath = sp.relative_to(dh_path.relative_path)
+                    if relpath != Path():
+                        res.add(generate_entry(relpath.parts[0], True))
+                except ValueError:
+                    continue
+        return res
+
+    def _api_listdir(self, path: DagshubPath, include_size: bool = False) -> Optional[List[ContentAPIEntry]]:
+        response, hit = self._check_listdir_cache(path.relative_path.as_posix(), include_size)
+        if hit:
+            return response
+        response = self.http_get(path.content_url,
+                                 params={'include_size': 'true'} if include_size else {},
+                                 headers=config.requests_headers)
+        if response.status_code >= 400:
+            logger.debug(f"Got HTTP code {response.status_code} while listing {path}, no results will be returned")
+            return None
+        res = []
+        for entry_raw in response.json():
+            entry = dacite.from_dict(ContentAPIEntry, entry_raw)
+            # Ignore storage root entries, we handle them separately in a different place
+            if entry.type == "storage":
+                continue
```

Comment:
This function is internal to the fs, and none of the consumers want to see the storage entry, since we're handling them separately.
If I start returning storage entries, I'll have to handle it in every consuming function.
The response is also cached, so this means that now the consumer functions need to do the redundant work of going through the entries every time.

Reply:
My argument here is that we are never gonna have to handle the storage functions in any other way than we're already handling them here

Reply:
Not critical.